worker/file_util.py

```diff
@@ -0,0 +1,187 @@
+    object after gunzipping.
+
+    Raises an IOError if the archive is not valid.
+    """
+    class UnGzipStream(object):
+        def __init__(self, fileobj):
+            self._fileobj = fileobj
+            self._decoder = zlib.decompressobj(16 + zlib.MAX_WBITS)
+            self._buffer = ''
+            self._finished = False
+    
+        def read(self, num_bytes=None):
+            # Read more data, if we need to.
+            while not self._finished and (num_bytes is None or len(self._buffer) < num_bytes):
+                chunk = self._fileobj.read(num_bytes) if num_bytes is not None else self._fileobj.read()
+                if chunk:
+                    self._buffer += self._decoder.decompress(chunk)
+                else:
+                    self._buffer += self._decoder.flush()
+                    self._finished = True
+            if num_bytes is None:
+                num_bytes = len(self._buffer)
+            result = self._buffer[:num_bytes]
+            self._buffer = self._buffer[num_bytes:]
+            return result
+
+        def close(self):
+            self._fileobj.close()
+    
+    # Note, that we don't use gzip.GzipFile or the gunzip shell command since
+    # they require the input file-like object to support either tell() or
+    # fileno(). Our version requires only read() and close().
+    return UnGzipStream(fileobj)
+
+
+def gzip_string(string):
+    """
+    Gzips the given string.
+    """
+    with closing(StringIO()) as output_fileobj:
+        with gzip.GzipFile(None, 'wb', 6, output_fileobj) as fileobj:
+            fileobj.write(string)
+        return output_fileobj.getvalue()
+
+
+def un_gzip_string(string):
+    """
+    Gunzips the given string.
+
+    Raises an IOError if the archive is not valid.
+    """
+    with closing(StringIO(string)) as input_fileobj:
+        with gzip.GzipFile(None, 'rb', fileobj=input_fileobj) as fileobj:
+            return fileobj.read()
+
+
+def read_file_section(file_path, offset, length):
+    """
+    Reads length bytes of the given file from the given offset.
+    """
+    file_size = os.stat(file_path).st_size
+    if offset >= file_size:
+        return ''
+    with open(file_path) as fileobj:
+        fileobj.seek(offset, os.SEEK_SET)
+        return fileobj.read(length)
+
+
+def summarize_file(file_path, num_head_lines, num_tail_lines, max_line_length, truncation_text):
+    """
+    Summarizes the file at the given path, returning a string containing the
+    given numbers of lines from beginning and end of the file. If the file needs
+    to be truncated, places truncation_text at the truncation point.
+    """
+    assert(num_head_lines > 0 or num_tail_lines > 0)
+
+    def ensure_ends_with_newline(lines):
+        if lines and not lines[-1].endswith('\n'):
+            lines[-1] += '\n'
+    
+    file_size = os.stat(file_path).st_size
+    with open(file_path) as fileobj:
```

Comment:
head_lines and tail_lines should end at clean line boundaries, but this seems like the last and first lines will be incomplete, respectively in this implementation